File: src/holmes_engine/engine.py
from typing import List, Dict

from holmes_rule.rule import HolmesRule
from .worker import Worker


class Engine:
    def __init__(self) -> None:
        # For results caching, fetch by application layer with Engine.fetch_results()
        self.results_cached_buffer = []
        # One worker is a engine instance correspondinng one rule
        self.workers:Dict[str, Worker] = {}
        # Record some tag should be checked by which rules(workers). The int-item is worker_ind
        self.tag_index:Dict[str, set[str]] = {}

    def add_holmes_rule(self, rule:HolmesRule):
        # check exist
        if rule.ruleid in self.workers:
            return "exist"
        # add worker
        worker = Worker(rule=rule, rulename=rule.ruleid)
        self.workers[rule.ruleid] = worker
        # add tag2worker index
        for tag_node in rule.tag_nodes:
            tag_id = tag_node.tag_rule.id
            if tag_id not in self.tag_index: self.tag_index[tag_id] = set()
            self.tag_index[tag_id].add(rule.ruleid)
        return "success"

    def del_holmes_rule(self, ruleid):
        # check exist
        if ruleid not in self.workers:
            return "not exist"
        # del worker
        del self.workers[ruleid]
        return "success"

    def update_holmes_rule(self, rule:HolmesRule):
        r = self.del_holmes_rule(ruleid=rule.ruleid)
        if r == "not exist": return r
        self.add_holmes_rule(rule=rule)
        return "success"

    def fetch_results(self):
        # will clean the cached results
        results = self.results_cached_buffer
        self.results_cached_buffer = []
        return results

    def process_event(self, event):
        tag_id = event["holmes-tag"]
        if tag_id not in self.tag_index: return
        for holmes_rule_id in self.tag_index[tag_id]:
            worker = self.workers[holmes_rule_id]
            results = worker.process_event(event=event)
            worker.prune(event_time=event["time"])
            if results != []:
                for result in results:
                    self.results_cached_buffer.append({"rulename": worker.rulename, "output": result})

File: java/com/amber/holmes/Engine.java
package com.amber.holmes;

import java.util.*;

public class Engine {
    // For results caching, fetch by application layer with Engine.fetchResults()
    private List<Map<String, Object>> resultsCachedBuffer;
    // One worker is an engine instance corresponding to one rule
    private Map<String, Worker> workers;
    // Record which tags should be checked by which rules(workers). The value is a
    // set of worker IDs
    private Map<String, Set<String>> tagIndex;

    public Engine() {
        this.resultsCachedBuffer = new ArrayList<>();
        this.workers = new HashMap<>();
        this.tagIndex = new HashMap<>();
    }

    public String addHolmesRule(HolmesRule rule) {
        // check exist
        if (workers.containsKey(rule.ruleId)) {
            return "exist";
        }
        // add worker
        Worker worker = new Worker(rule, rule.ruleId);
        workers.put(rule.ruleId, worker);
        // add tag2worker index
        for (TagNode tagNode : rule.tagNodes) {
            String tagId = tagNode.tagRule.id;
            tagIndex.computeIfAbsent(tagId, k -> new HashSet<>()).add(rule.ruleId);
        }
        return "success";
    }

    public String delHolmesRule(String ruleId) {
        // check exist
        if (!workers.containsKey(ruleId)) {
            return "not exist";
        }
        // del worker
        workers.remove(ruleId);
        return "success";
    }

    public String updateHolmesRule(HolmesRule rule) {
        String result = delHolmesRule(rule.ruleId);
        if (result.equals("not exist"))
            return result;
        return addHolmesRule(rule);
    }

    public List<Map<String, Object>> fetchResults() {
        // will clean the cached results
        List<Map<String, Object>> results = new ArrayList<>(resultsCachedBuffer);
        resultsCachedBuffer.clear();
        return results;
    }

    public void processEvent(Map<String, Object> event) {
        String tagId = (String) event.get("holmes-tag");
        if (!tagIndex.containsKey(tagId))
            return;
        for (String holmesRuleId : tagIndex.get(tagId)) {
            Worker worker = workers.get(holmesRuleId);
            List<List<Object>> results = worker.processEvent(event);
            worker.prune((long) (Integer) event.get("time"));
            if (!results.isEmpty()) {
                for (List<Object> result : results) {
                    Map<String, Object> output = new HashMap<>();
                    output.put("rulename", worker.rulename);
                    output.put("output", result);
                    resultsCachedBuffer.add(output);
                }
            }
        }
    }
}
